torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -254,12 +272,37 @@ def init_all_gather_outputs(
 
     def init_unsharded_param(self):
         if hasattr(self, "_unsharded_param"):
-            return  # already initialized
-        # For the default path (no post-all-gather), the all-gather output
-        # gives the unsharded parameter data directly
-        assert len(self.all_gather_outputs) == 1
+            if not self._use_all_gather_extensions:
+                return  # already initialized
+            for tensor in self._unsharded_inner_tensors:
+                if (storage := tensor.untyped_storage()).size() == 0:
```

Comment:
Calls to `untyped_storage()` and `resize()` can be considered expensive for CPU overhead. This is not ideal in terms of readability, but we prefer to call `untyped_storage()` just once.